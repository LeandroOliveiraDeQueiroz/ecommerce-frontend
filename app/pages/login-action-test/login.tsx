import { Form } from "react-router";
import { getStyledButton } from "~/components/styledButton/styledButton";


const StyledButton = getStyledButton("blue-500", "white");


export function Login() {

    return (
        <div className="flex-1 flex flex-col justify-center items-center">
            <div className="p-4 border-2 rounded-md border-gray-500">
                <h1 className="text-2xl m-auto mt-3 mb-3">Login</h1>
                <Form method="post">
                    <div>
                        <label>Email</label>
                        <input name="email" type="email" />
                    </div>
                    <div>
                        <label>Senha</label>
                        <input name="password" type="password" />
                    </div>
                    <div className="text-center">
                        <StyledButton type="submit">Logar</StyledButton>
                    </div>
                </Form>
            </div>
        </div>
    )
}