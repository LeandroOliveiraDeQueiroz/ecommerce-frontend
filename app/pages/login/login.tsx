import { useState } from "react";
import { useNavigate } from "react-router";
import { getStyledButton } from "~/components/styledButton/styledButton";
import { useAuthContext } from "~/contexts/auth/auth";
import authService from "~/services/auth/auth";


const StyledButton = getStyledButton("blue-500", "white");


export function Login() {
    const navigate = useNavigate();
    const { setLoggedUser, isLogged } = useAuthContext();
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");

    const handleLogin = async () => {
        try {
            const logged = await authService.login({ email, password });
            console.log("logged", logged);
            //TODO setLogged context

            if (logged) {
                setLoggedUser({
                    name: logged.name, accessToken: logged.accessToken
                })

                navigate('/');
            }

        } catch (error) {
            console.log("error:", error);
        }
    }

    const handleChangeEmail = (value: string) => {
        setEmail(value)
    }

    const handleChangePassword = (value: string) => {
        setPassword(value)
    }

    if (isLogged)
        navigate("/")

    return (
        <div className="flex-1 flex flex-col justify-center items-center">
            <div className="p-4 border-2 rounded-md border-gray-500">
                <h1 className="text-2xl m-auto mt-3 mb-3">Login</h1>
                <div>
                    <div>
                        <label>Email</label>
                        <input value={email} onChange={(ev) => { handleChangeEmail(ev.target.value) }} type="email" />
                    </div>
                    <div>
                        <label>Senha</label>
                        <input value={password} onChange={(ev) => { handleChangePassword(ev.target.value) }} type="password" />
                    </div>
                    <div className="text-center">
                        <StyledButton onClick={handleLogin}>Logar</StyledButton>
                    </div>
                </div>
            </div>
        </div>
    )
}