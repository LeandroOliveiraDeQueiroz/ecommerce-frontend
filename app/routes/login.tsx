import { Login } from "~/pages/login/login";
import type { Route } from "./+types/home";
import authService from "~/services/auth/auth";
import { Form, useNavigate } from "react-router";
import React, { useEffect } from "react";
import { useAuthContext } from "~/contexts/auth/auth";
import { useSnackbar } from "notistack";
import type { IFavoriteProductList } from "~/types";
import favoriteProductsListService from "~/services/favoriteProductsList/favoriteProductsList";
import { useFavoriteProductListContext } from "~/contexts/favoriteProductsList/favoriteProductsList";
import * as yup from 'yup'

export function meta({ }: Route.MetaArgs) {
  return [
    { title: "Luizalabs login" },
    { name: "Login", content: "Pagina de login" },
  ];
}

const schema = yup.object().shape({
  email: yup
    .string()
    .required("Email é obrigatório")
    .email("Email não válido"),
  password: yup
    .string()
    .required('Senha é obrigatório')
});


export async function clientAction({
  request,
}: Route.ClientActionArgs) {
  let formData = await request.formData();
  let email = formData.get("email");
  let password = formData.get("password");

  try {

    console.log(email, password)

    await schema.validate({ email, password });

    const userData = await authService.login({ email, password });

    if (!userData) {
      return { error: true }
    }

    const list: IFavoriteProductList | null = await favoriteProductsListService.read({ accessToken: userData?.accessToken });

    return { success: true, userData: userData, redirectTo: '/', list: list }
  } catch (error) {

    if (error instanceof yup.ValidationError) {
      return { error: true, message: error.message }
    }

    return { error: true, message: "Email ou senha incorreta" }

  }
}


export default function LoginPage({
  actionData,
}: Route.ComponentProps) {
  const navigate = useNavigate();
  const { login, isAuthenticated } = useAuthContext();
  const { get } = useFavoriteProductListContext();
  const { enqueueSnackbar } = useSnackbar();

  useEffect(() => {
    if (isAuthenticated && !actionData?.success) {
      enqueueSnackbar('Usuário já logado', { variant: 'success' });
      navigate("/");
    }
  }, [isAuthenticated])

  useEffect(() => {
    if (actionData?.success) {
      enqueueSnackbar('Login realizado com sucesso!', { variant: 'success' });
      login(actionData.userData)
      get(actionData.list as IFavoriteProductList);
      navigate(actionData.redirectTo);
    } else if (actionData?.error) {
      enqueueSnackbar(actionData.message || "Erro inesperado", { variant: 'error' });
      console.error("Error:", actionData?.error)
    }
  }, [actionData, navigate]);


  if (actionData) {
    console.log("action", actionData)
  }

  return (
    <Login />
  )
}
