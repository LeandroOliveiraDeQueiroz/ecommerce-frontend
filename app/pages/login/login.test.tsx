import { render } from '@testing-library/react'
import { Login } from './login'
import { createRoutesStub } from 'react-router';

it('should render correctly and matches snapshot', () => {

    const Stub = createRoutesStub([
        {
            path: "/login",
            Component: Login,
        }
    ]);

    const { container } = render(<Stub initialEntries={["/login"]} />);

    expect(container).toMatchSnapshot();
});