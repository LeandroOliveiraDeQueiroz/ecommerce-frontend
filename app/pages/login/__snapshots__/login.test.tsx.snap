// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly and matches snapshot 1`] = `
<div>
  <div
    class="flex-1 flex flex-col justify-center items-center"
  >
    <div
      class="p-4 border-2 rounded-md border-gray-500"
    >
      <h1
        class="text-2xl m-auto mt-3 mb-3 text-center"
      >
        Login
      </h1>
      <form
        action="/login"
        data-discover="true"
        method="post"
      >
        <div>
          <label
            class="block mb-1  text-gray-900 dark:text-white"
          >
            Email
          </label>
          <input
            class="mb-5 bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            name="email"
            type="email"
          />
        </div>
        <div>
          <label
            class="block mb-1  text-gray-900 dark:text-white"
          >
            Senha
          </label>
          <input
            class="mb-5 bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            name="password"
            type="password"
          />
        </div>
        <div
          class="text-center"
        >
          <button
            class="border-blue-400 bg-white text-blue-400 border-2 rounded-md px-4 py-2 transition-colors duration-300 hover:bg-blue-400 hover:text-white"
            type="submit"
          >
            Logar
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
